draft-ietf-mls-architecture.md

```diff
@@ -760,8 +760,14 @@ they don't receive another Commit first.
 The Delivery Service can rely on the `epoch` and `content_type` fields of an
 MLSMessage for providing an order only to handshake messages, and possibly even
 filter or reject redundant Commit messages proactively to prevent them from
-being broadcast. Alternatively, the Delivery Service could simply apply an order
-to all messages and rely on clients to ignore redundant Commits.
+being broadcast. However, this is not to say that the Delivery Service should
+enforce an absolutist view of the current group `epoch` and reject all messages
+that do not match it. Delivery Services have a significantly limited view of the
+group state, and of which messages will be acceptable to clients. Once a
+Delivery Service is confident that a given commit was processed by a client, if
+the client continues to send messages in the previous epoch, the Delivery
+Service should assume the commit was silently rejected and stop
+blocking/filtering.
```

Comment:
This last sentence is speculative and very specific to a particular architectural choice. Please delete it.

```suggestion
group state, and of which messages will be acceptable to clients.
```

Reply:
If the purpose of the PR is to remove speculative language about what a DS might do, the cure here is worse than the disease and introduces further speculative text.

Reply:
The purpose of this PR is to say that the DS may filter commits, in the interest of efficiency / etc, but can't be the ultimate arbiter of which commit is correct and needs to defer to clients eventually. The last sentence was added to make the deference point more explicit after ekr's comment: https://github.com/mlswg/mls-architecture/pull/257#issuecomment-2151042899

Reply:
But you are making a very specific comment about what the DS should assume which I believe is rarely correct and has substantial DoS attack consequences. It introduces a whole new attack surface.

Reply:
The DoS attack is documented in the parent issue #251, i.e., when the DS tries to enforce an incorrect view of the epoch. The alternate approaches that you bring up in #251 are not secure:

- Asking the DS to rollback an invalid commit. As you say there, there's no way to prevent this from being abused. Proving to the DS that a commit is invalid would require revealing secret keys in a way that would impair group security. Trying to guard with a quorum is also vulnerable to sybil.
- Members do an external join. This still has DoS potential, and allows the DS / an attacker to make arbitrary changes to the ratchet tree & group state, which members will then accept uncritically.

The only way to mitigate DoS and preserve group security is for DS to defer to clients

Reply:
I'm going to flatly disagree with your last sentence "the only way to mitigate DoS" and your proposed solution in the original PR. Your proposed solution assumes that the clients act correctly when the only possible way they could have gotten into the situation in the first place was a client bug or a malicious client. Rolling back to an earlier commit may in fact be the attacker's intended form of service denial. 

Let's be clear here that the DoS attack you are talking about is in practice limited to members (because in many cases the DS/AS knows the client credential that matches that member). It is completely reasonable in some environments that a DS "fix" a misbehaving client, by banishing the offending client, not by letting the client further run afoul of policy rules.

Reply:
I think you're confused about what I'm proposing here. Deference means that the DS allows clients to behave in ways that are potentially incorrect, and allowing honest clients to decide to ignore messages from misbehaving clients. Is there a specific example you could walk through, where deference doesn't work?

Yes, the DoS attack is limited to members. But it's not correct to say that the DS has any ability to identify which client was misbehaving. The DS isn't able to identify incorrect MACs, UpdatePaths, etc. You could try guarding this mechanism with a quorum vote, but again, that's vulnerable to sybil attacks.